packages/babel-plugin-optimize-react/README.md

```diff
@@ -0,0 +1,58 @@
+# babel-plugin-optimize-react
+
+This Babel 7 plugin optimizes React hooks by transforming common patterns into more effecient output when using with tools such as [Create React App](https://github.com/facebook/create-react-app). For example, with this plugin the following output is optimized as shown:
+
+```js
+// Original
+var _useState = Object(react__WEBPACK_IMPORTED_MODULE_1_["useState"])(Math.random()),
+    _State2 = Object(_Users_gaearon_p_create_rreact_app_node_modules_babel_runtime_helpers_esm_sliceToArray_WEBPACK_IMPORTED_MODULE_0__["default"])(_useState, 1),
+    value = _useState2[0];
+    
+// With this plugin
+var useState = react__WEBPACK_IMPORTED_MODULE_1_.useState;
+var __ref__0 = useState(Math.random());
+var value = __ref__0[0];
+```
+
+## Named imports to hooks get transformed
+
+```js
+// Original
+import React, {useState} from 'react';
+
+// With this plugin
+import React from 'react';
+const {useState} = React;
+```
+
+## Array destructuring transform for React's built-in hooks
```

Comment:
This should be configurable or ideally communicate with targets/browserslists/`@babel/preset-env`. Array [spread syntax is supported by all evergreen browsers](https://kangax.github.io/compat-table/es6/) and that for quite some time now. The optimization is only interesting if you need to support IE11. See [example preset-env without IE11](https://babeljs.io/repl#?babili=false&browsers=%3E%201%25%2C%20not%20IE%2011&build=&builtIns=false&spec=true&loose=false&code_lz=GYVwdgxgLglg9mABAYTgWwA4IKZigCgwCc4MBnASkQG8AoASAgTKkQG0WBDKbAGkTLYoAZSjdsAXUQBeRACVsnaADoQg0ePwUA3LQC-QA&debug=false&forceAllTransforms=false&shippedProposals=false&circleciRepo=&evaluate=true&fileSize=false&timeTravel=false&sourceType=script&lineWrap=true&presets=env&prettier=true&targets=Node-8.0&version=7.4.5&externalPlugins=)

Reply:
We opted to do this because the runtime performance of spread syntax was considerably slower in all browsers when I tests this compared to the transformed version. This might have changed since, as this plugin was created 7 months ago.

Reply:
Makes sense. Not sure what the current status of destructuring optimizations in v8 but [they had plans to improve it](https://docs.google.com/document/d/1hWb-lQW4NSG9yRpyyiAA_9Ktytd5lypLnVLhPX9vamE/edit) a few months ago. Do you have some numbers about the performance gain?